File: src/backend/calculator.js
function parseExpression(expr) {
    const outputQueue = [];
    const operatorStack = [];
  
    for (let i = 0; i < expr.length; i++) {
      let char = expr[i];
  
      if (/\s/.test(char)) {
        continue;
      }
  
      if (/\d/.test(char)) {
        let numStr = char;
        while (i + 1 < expr.length && /\d/.test(expr[i + 1])) {
          numStr += expr[++i];
        }
        outputQueue.push(parseFloat(numStr));
      }
  
      else if (/[\+\-\*\/]/.test(char)) {
        let precedence = getPrecedence(char);
  
        while (operatorStack.length > 0 && precedence <= getPrecedence(operatorStack[operatorStack.length - 1])) {
          outputQueue.push(operatorStack.pop());
        }
  
        operatorStack.push(char);
      }
  
      else if (char === '(') {
        operatorStack.push(char);
      }
  
      else if (char === ')') {
        while (operatorStack.length > 0 && operatorStack[operatorStack.length - 1] !== '(') {
          outputQueue.push(operatorStack.pop());
        }
  
        if (operatorStack.length === 0 || operatorStack[operatorStack.length - 1] !== '(') {
          throw new Error('Tanda kurung tidak cocok');
        }
  
        operatorStack.pop();
      }
  
      else {
        throw new Error(`Karakter tidak valid: ${char}`);
      }
    }
  
    while (operatorStack.length > 0) {
      let operator = operatorStack.pop();
  
      if (!/[\+\-\*\/]/.test(operator)) {
        throw new Error(`Operator tidak valid: ${operator}`);
      }
  
      outputQueue.push(operator);
    }
  
    let stack = [];
    for (let i = 0; i < outputQueue.length; i++) {
      let item = outputQueue[i];
  
      if (typeof item === 'number') {
        stack.push(item);
      } else if (/[\+\-\*\/]/.test(item)) {
        let b = stack.pop();
        let a = stack.pop();
  
        if (item === '/' && b === 0) {
          throw new Error('Pembagian dengan 0');
        }
  
        let result = evaluateExpression(a, b, item);
        stack.push(result);
      }
    }
  
    if (stack.length !== 1) {
      throw new Error('Tidak bisa dihitung');
    }
  
    return stack.pop();
}
  
function getPrecedence(operator) {
    switch (operator) {
      case '+':
      case '-':
        return 1;
      case '*':
      case '/':
        return 2;
      default:
        return 0;
    }
}
  
function evaluateExpression(a, b, operator) {
    switch (operator) {
      case '+': return a + b;
      case '-': return a - b;
      case '*': return a * b;
      case '/': return a / b;
    }
}

function main(){
    const inputExpr = '4 *';

    try 
    {
      const result = parseExpression(inputExpr);
      console.log(`${inputExpr} = ${result}`);
    } 
    catch (error) 
    {
      console.error(`Error: ${error.message}`);
    }
}

// main();
export default parseExpression;

File: src/backend/calculator.go
package main

import (
    "fmt"
    "strconv"
    "regexp"
    "unicode"
    "math"
)

func parseExpression(expr string) (result float64, err error) {
    var outputQueue []float64
    var operatorStack []string

    for i := 0; i < len(expr); i++ {
        char := string(expr[i])

        if char == " " {
            continue
        }

        if _, err := strconv.Atoi(char); err == nil {
            numStr := char
            for i+1 < len(expr) && unicode.IsDigit(rune(expr[i+1])) {
                numStr += string(expr[i+1])
                i++
            }
            num, _ := strconv.ParseFloat(numStr, 64)
            outputQueue = append(outputQueue, num)
        } else if regexp.MustCompile(`[+\-*/^]`).MatchString(char) {
            precedence := getPrecedence(char)

            for len(operatorStack) > 0 && precedence <= getPrecedence(operatorStack[len(operatorStack)-1]) {
                result := evaluateExpression(outputQueue[len(outputQueue)-2], outputQueue[len(outputQueue)-1], operatorStack[len(operatorStack)-1])
                outputQueue = outputQueue[:len(outputQueue)-2]
                outputQueue = append(outputQueue, result)
                operatorStack = operatorStack[:len(operatorStack)-1]
            }

            operatorStack = append(operatorStack, char)
        } else if char == "(" {
            operatorStack = append(operatorStack, char)
        } else if char == ")" {
            for len(operatorStack) > 0 && operatorStack[len(operatorStack)-1] != "(" {
                result := evaluateExpression(outputQueue[len(outputQueue)-2], outputQueue[len(outputQueue)-1], operatorStack[len(operatorStack)-1])
                outputQueue = outputQueue[:len(outputQueue)-2]
                outputQueue = append(outputQueue, result)
                operatorStack = operatorStack[:len(operatorStack)-1]
            }

            if len(operatorStack) == 0 || operatorStack[len(operatorStack)-1] != "(" {
                panic("Tanda kurung tidak cocok")
            }

            operatorStack = operatorStack[:len(operatorStack)-1]
        } else {
            panic(fmt.Sprintf("Karakter tidak valid: %s", char))
        }
    }

    for len(operatorStack) > 0 {
        operator := operatorStack[len(operatorStack)-1]
        operatorStack = operatorStack[:len(operatorStack)-1]

        if !regexp.MustCompile(`[+\-*/^]`).MatchString(operator) {
            panic(fmt.Sprintf("Operator tidak valid: %s", operator))
        }

        b := outputQueue[len(outputQueue)-1]
        outputQueue = outputQueue[:len(outputQueue)-1]
        a := outputQueue[len(outputQueue)-1]
        outputQueue = outputQueue[:len(outputQueue)-1]

        if operator == "/" && b == 0 {
            panic("Pembagian dengan 0")
        }

        result := evaluateExpression(a, b, operator)
        outputQueue = append(outputQueue, result)
    }

    if len(outputQueue) != 1 {
        // fmt.Print(outputQueue)
        panic("Tidak bisa dihitung")
    }

    return outputQueue[0], nil
}

func getPrecedence(operator string) int {
    switch operator {
    case "+", "-":
        return 1
    case "*", "/":
        return 2
    case "^":
        return 3
    default:
        return 0
    }
}

func evaluateExpression(a float64, b float64, operator string) float64 {
    switch operator {
    case "+":
        return a + b
    case "-":
        return a - b
    case "*":
        return a * b
    case "/":
        return a / b
    case "^":
        return math.Pow(a, b)
    default:
        panic(fmt.Sprintf("Operator tidak valid: %s", operator))
    }
}

func main() {
    var result float64
    var err error
    inputExpr := "-5+3"

    result, err = parseExpression(inputExpr)
    if err != nil {
        fmt.Printf("Error: %v\n", err)
    } else {
        fmt.Printf("Result: %f\n", result)
    }

    
}
